xocto/fields/ranges.py

```diff
@@ -0,0 +1,68 @@
+"""
+Adapter fields for postgres range types that return xocto.Range python objects.
+"""
+import datetime
+import json
+from typing import Optional
+
+from django.contrib.postgres import fields as pg_fields
+from django.contrib.postgres.fields import ranges as pg_ranges
+from django.contrib.postgres.fields import utils as pg_utils
+from django.db import models
+
+from xocto import ranges
+
+
+class FiniteDateRangeField(pg_fields.DateRangeField):
+    """
+    A DateRangeField with Inclusive-Inclusive [] bounds.
+
+    The underlying postgres type will always store the range as [). This field
+    will always translate to [].
+
+    Accepts and returns xocto.ranges.FiniteDateRange objects.
+    """
+
+    def get_prep_value(
+        self, value: Optional[ranges.FiniteDateRange]
+    ) -> Optional[ranges.FiniteDateRange]:
+        if value is None:
+            return None
+        return pg_ranges.DateRange(lower=value.start, upper=value.end, bounds="[]")
+
+    def from_db_value(
+        self,
+        value: Optional[pg_ranges.DateRange],
+        expression: object,
+        connection: object,
+    ) -> Optional[ranges.FiniteDateRange]:
+        if value is None:
+            return None
+        # DateRange is always stored as [) in the database, so we need to remove
+        # a day from the upper bound to make it inclusive.
+        return ranges.FiniteDateRange(
+            start=value.lower, end=value.upper - datetime.timedelta(days=1)
+        )
+
+    def to_python(self, value: Optional[str]) -> Optional[ranges.FiniteDateRange]:
+        if value is None:
+            return None
+        obj = json.loads(value)
+        return ranges.FiniteDateRange(
+            start=self.base_field.to_python(obj["start"]),
+            end=self.base_field.to_python(obj["end"]),
+        )
+
+    def value_to_string(self, obj: models.Model) -> Optional[str]:
+        value: Optional[ranges.FiniteDateRange] = self.value_from_object(obj)
+        if value is None:
+            return None
+        base_field = self.base_field
+        start = pg_utils.AttributeSetter(base_field.attname, value.start)
+        end = pg_utils.AttributeSetter(base_field.attname, value.end)
```

Comment:
I'm unsure what these two lines are doing

Reply:
They are partially cargo-culted from the underlying DateRange type. Initially I left them out, but `base_field` (DateTimeField) calls `value_from_object` which essentially does `getattr(self, attname)` (yay descriptors). This sets up the correct attributes so that the `getattr` works.